### ISSUE TITLE
Emit a warning when `sort` is used for range queries

### ISSUE THREAD

**Issue by charleskorn** [prometheus/prometheus#14115]
### Proposal

Using `sort` with range queries will not work: the series in a range query result will always be sorted by the series labels.

This may be confusing for users. It would be nice to emit a warning annotation when `sort` is used for a range query to explain what is happening.

(acknowledgement: this is @bboreham's idea from https://github.com/prometheus/prometheus/pull/14083#issuecomment-2112022774)

**Comment by bboreham** [comment]
It can be expensive on CPU to sort at every step of a range-query; we could save that overhead.

[sort_by_label](https://prometheus.io/docs/prometheus/latest/querying/functions/#sort_by_label) could work for range queries, but Prometheus currently overrides that too.

**Comment by epochstamp** [comment]
@bboreham I would be interested by the ability to sort range queries (it should be possible via the value-based sorting function without sorting for every steps, we could add an extra-parameter to select order criterion such as last value or area under the curve...)

Sorry for the silly question but I am not yet familiar with Prometheus architecture : do you mean that, once the query executed and series are gathered, Prometheus re-sort the series ? If so, could this overriding be disabled conditionnally on the presence of the sorting functions in the labels ? (to limit overhead it could be restricted to range queries having the sorting function at the top of the query tree, i.e. `sort_by_label(`<your_promql_query_to_be_sorted>`)

**Comment by bboreham** [comment]
> do you mean that, once the query executed and series are gathered, Prometheus re-sort the series

Yes.

> could this overriding be "overrided" again conditionnally on the presence of the sorting functions in the labels ? 

Yes, technically.

"sort by last value" or "sort by area" would need to be another proposal.  

Note that many comparable proposals have been rejected on the grounds that it's best implemented in the client (e.g. Grafana).

**Comment by epochstamp** [comment]
"Note that many comparable proposals have been rejected on the grounds that it's best implemented in the client (e.g. Grafana)."

I also saw this argument multiple times. Either I am constantly missing something, or this does not hold for queries that do not have a fixed output cardinality. Moreover, the reverse argument is also held in Grafana's side to reject sorting features, (e.g. [here](https://github.com/grafana/grafana/issues/43537) where they suggest to either use a Grafana transform [which doesn't work for the above-mentioned use case here]) or **sorting the result via the query itself**.

While I agree that it would be way more convenient to implement range sorting on the client's side, it could also be worse to evaluate how such a feature would be difficult to implement from PromQL side (especially in regards of the Prometheus ordering override).

**Comment by bboreham** [comment]
> it could also be worse

Is that a typo for "worth it" ?

I do encourage you to make your own proposal, and to include all the background information.  This issue will be closed when a warning is added, so the idea here would then be hard to discover.

**Comment by epochstamp** [comment]
It was indeed a typo >< I just edited my comment.

the proposal is on my todo list :) thanks for your clarifications

**Comment by alexgreenbank** [comment]
Hello from the bug scrub.

Note that the majority of the discussion above is about a separate issue.

This issue remains the addition of an annotation when sorting is used for range queries, which should be a relatively simple bit of work.

**Comment by poi1649** [comment]
Hi all, I just submitted a PR for this. Would appreciate it if you could take a look and let me know if I understood the issue correctly and implemented it as expected. :)

**Comment by Gokula-6sg** [comment]
Hi! 👋 I’d like to work on this issue. Could you please confirm if it’s still open for contribution?

**Comment by adity-a34** [comment]
Hi @poi1649 
Your PR seems to be marked stale, are you still trying to implement this issue

@beorn7 I would like to implement solution to the issue, & If it is open for contribution
please assign it to me.
Thank you.

**Comment by beorn7** [comment]
@poi1649's PR only got a review 3 weeks ago after a long waiting time. I would give @poi1649 a bit more time to react.

The [PR](https://github.com/prometheus/prometheus/pull/16628) seems to be almost there, according to the review so far. So the remaining work here would simply be to complete the work already done. This might be so little that it could be done by the reviewer themself rather than finding a new contributor to work on it.

**Comment by poi1649** [comment]
@beorn7 Thanks for waiting on this, I’ve addressed the comments, so feel free to take another look when you get a chance.
